File: binary_tree/104.二叉树的最大深度.go
/*
 * @lc app=leetcode.cn id=104 lang=golang
 *
 * [104] 二叉树的最大深度
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
// func maxDepth(root *TreeNode) int {
// 	if root == nil {
// 		return 0
// 	}
// 	queue := make([]*TreeNode, 0)
// 	queue = append(queue, root)
// 	depth := 0
// 	for len(queue) > 0 {
// 		size := len(queue)
// 		for i := 0; i < size; i++ {
// 			node := queue[0]
// 			queue = queue[1:]
// 			if node.Left != nil {
// 				queue = append(queue, node.Left)
// 			}
// 			if node.Right != nil {
// 				queue = append(queue, node.Right)
// 			}
// 		}
// 		depth++
// 	}
// 	return depth
// }
func maxDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	res := 1
	return getDepth(root, 1, res)

}

func getDepth(root *TreeNode, depth int, res int) int {
	if depth > res {
		res = depth
	}
	if root.Left == nil && root.Right == nil {
		return res
	}
	if root.Left != nil {
		depth++
		res = getDepth(root.Left, depth, res)
		depth--
	}
	if root.Right != nil {
		depth++
		res = getDepth(root.Right, depth, res)
		depth--
	}
	return res
}

// @lc code=end

File: binary_tree/104.二叉树的最大深度.py
#
# @lc app=leetcode.cn id=104 lang=python3
#
# [104] 二叉树的最大深度
#


# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        res = 0

        def getDepth(node: TreeNode, depth: int):
            nonlocal res
            res = res if res > depth else depth
            if not node.left and not node.right:
                return
            if node.left:
                depth += 1
                getDepth(node.left, depth)
                depth -= 1  # 回朔
            if node.right:
                depth += 1
                getDepth(node.right, depth)
                depth -= 1  # 回朔
            return

        if not root:
            return res
        getDepth(root, 1)
        return res


# @lc code=end
